pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,207 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Generator, Iterator, List, Optional, Union
+
+import torch
+from torch.utils.data import BatchSampler, get_worker_info, Sampler
+from torch.utils.data.dataloader import DataLoader, IterableDataset
+
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+
+
+class FastForwardSampler:
```

Comment:
should we derive it from torch.utils.data.Sampler as well?

Reply:
Done !